crates/db_connection_pool/src/dbconnection/snowflakeconn.rs

```diff
@@ -0,0 +1,127 @@
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::sql::TableReference;
+use futures::stream;
+use snafu::prelude::*;
+use snowflake_api::SnowflakeApi;
+
+use super::AsyncDbConnection;
+use super::DbConnection;
+use super::Result;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Not implemented"))]
+    NotImplemented,
+
+    #[snafu(display("Unable to retrieve schema"))]
+    UnableToRetrieveSchema,
+
+    #[snafu(display("Unexpected query response, expected Arrow, got JSON"))]
+    UnexpectedResponse,
+
+    #[snafu(display("Error executing query: {source}"))]
+    SnowflakeQueryError {
+        source: snowflake_api::SnowflakeApiError,
+    },
+}
+
+pub struct SnowflakeConnection {
+    pub api: Arc<SnowflakeApi>,
+}
+
+impl<'a> DbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConnection {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn as_async(&self) -> Option<&dyn super::AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)>> {
+        Some(self)
+    }
+}
+
+#[async_trait]
+impl<'a> AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConnection {
+    fn new(api: Arc<SnowflakeApi>) -> Self {
+        SnowflakeConnection { api }
+    }
+
+    async fn get_schema(&self, table_reference: &TableReference) -> Result<SchemaRef> {
+        let table = table_reference.to_quoted_string();
+
+        let res = self
+            .api
+            .exec(format!("SELECT * FROM {table} limit 1").as_str())
+            .await
+            .context(SnowflakeQuerySnafu)?;
+
+        match res {
+            snowflake_api::QueryResult::Arrow(record_batches) => {
+                let schema = record_batches[0].schema();
+                return Ok(Arc::clone(&schema));
+            }
+            _ => UnableToRetrieveSchemaSnafu.fail()?,
+        }
+    }
+
+    async fn query_arrow(
+        &self,
+        sql: &str,
+        _: &[&'a (dyn Sync)],
+    ) -> Result<SendableRecordBatchStream> {
+        let sql = sql.to_string();
+
+        let res = self.api.exec(&sql).await.context(SnowflakeQuerySnafu)?;
+
+        match res {
+            snowflake_api::QueryResult::Arrow(record_batches) => {
+                let schema = record_batches[0].schema();
+                return Ok(Box::pin(MemoryStream::try_new(
+                    record_batches,
+                    schema,
+                    None,
+                )?));
+            }
+            snowflake_api::QueryResult::Empty => {
+                return Ok(Box::pin(RecordBatchStreamAdapter::new(
+                    Arc::new(Schema::empty()),
+                    stream::empty(),
+                )));
+            }
+            snowflake_api::QueryResult::Json(json) => {
+                tracing::debug!("Unexpected response from Snowflake query: {json}");
+                UnexpectedResponseSnafu.fail()?
```

Comment:
passing json into the UnexpectedResponseSnafu so it has a more detailed error message rather than just debug!

Reply:
Updated, thank you!